bip-0352.mediawiki

```diff
@@ -0,0 +1,457 @@
+* Inputs used to derive the shared secret are from the ''[[#inputs-for-shared-secret-derivation|Inputs For Shared Secret Derivation]]'' list
+
+=== Scanning transactions ===
+
+A transaction is a ''Silent Payments v0'' transaction and MUST be scanned if and only if all of the following are true:
+
+* The transaction contains at least one BIP341 taproot output
+* The transaction has at least one input from the ''[[#inputs-for-shared-secret-derivation|Inputs For Shared Secret Derivation]]'' list
+* The transaction does not contain a new, undefined output type (e.g. SegWit versions > 1)<ref name="skip_txs_with_unknown_prevouts">'''Why skip transactions that spend unknown output scripts?''' Skipping transactions that spend unknown output scripts allows us to have a clean upgrade path for Silent Payments by avoiding the need to scan the same transaction multiple times with different rule sets. If a fancy new output type is added in the future and Silent Payments v1 is released with support, we would want to avoid having to first scan the transaction with the silent payment v0 rules and then again with the silent payment v1 rules.</ref>
+* The transaction ''nVersion'' is < 4
+
+Otherwise, skip the transaction. This is to ensure forward compatibility with future versions of silent payments without requiring future versions to scan a transaction multiple times with different rule sets.
+
+=== Address encoding ===
+
+A silent payment address is constructed in the following manner:
+
+* Let ''B<sub>scan</sub>, b<sub>scan</sub> = Receiver's scan public key and corresponding private key''
+* Let ''B<sub>spend</sub>, b<sub>spend</sub> = Receiver's spend public key and corresponding private key''
+* Let ''B<sub>m</sub> = B<sub>spend</sub> + m·G'', where ''m'' an optional integer tweak for labeling
+** In the case of ''m'' = 0, no label is applied and ''B<sub>m</sub> = B<sub>spend</sub>''
+* The final address is a [https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki Bech32m] encoding of:
+** The human-readable part "sp" for mainnet, "tsp" for testnets (e.g.  signet, testnet)
+** The data-part values:
+*** The character "q", to represent a silent payment address of version 0
+*** The 66 byte concatenation of the receiver's public keys, ''ser<sub>P</sub>(B<sub>scan</sub>) || ser<sub>P</sub>(B<sub>m</sub>)''
+
+Note: [https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki BIP173] imposes a 90 character limit for Bech32 strings, whereas a silent payment address requires 117 characters.
```

Comment:
I was wondering how exactly we end up with a limit of 117 characters, maybe it's worth it to include a short breakdown (e.g. as a foot note)? IIUC, this should be
- HRP [2-3 characters]
- separator [1 character]
- version [1 character]
- payload, 66 bytes concatenated pubkeys [ ceil(66*8/5) = 106 characters]
- checksum [6 characters]

i.e. silent payment addresses have a fixed size of 117 characters on testnet/signet/regtest and 116 characters on mainnet.

Reply:
I've updated this to recommend raising the character limit to 1023 for silent payment addresses. This is because new versions of silent payments may add additional bytes to the data field. For forward compatibility, the specification says a silent payments v0 sender can send to v0 - v30 addresses - they just read the first 66 bytes and discard the rest. This doesn't work if there is a size limit on the address since a v0 sender would fail to decode a v1 - v30 address.

Also, raising the limit seems "safe" per BIP173 and IIUC this is what lightning is using for their invoices.

Reply:
Also, yes, it would be nice to add a breakdown to the footnotes! Thanks for providing one!